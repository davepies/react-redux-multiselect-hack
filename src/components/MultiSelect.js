import React, { Component } from 'react'
import Radium from 'radium'
import onClickOutside from 'react-onclickoutside'

class MultiSelect extends Component {

  state = {
    shouldShowValues: false
  }

  handleFocus = () => {
    this.setState({ shouldShowValues: true })
  }

  handleClickOutside = () => {
    this.setState({ shouldShowValues: false })
  }

  render () {
    const {
      onFilterChange,
      onValueSelect,
      onValueRemove,
      values = [],
      selectedValues = []
    } = this.props

    const {
      shouldShowValues,
    } = this.state

    return (
      <div style={styles.wrapper}>
        <input
          type='text'
          onFocus={() => this.handleFocus()}
          onChange={(e) => onFilterChange(e.target.value)}
        />
        <div style={styles.values}>
          <ul style={[styles.list, styles[shouldShowValues ? 'listVisible' : null]]}>
            {values.map(
              (value, i) => (
                <li key={i}>
                  <span
                    style={styles.listItem}
                    onClick={() => { onValueSelect(value) }}
                  >
                    {value}
                  </span>
                </li>
              )
            )}
          </ul>
          <span style={styles.darth} />
        </div>
        {!!selectedValues.length &&
          <div>
            <span>You have selected:</span>
            {
              selectedValues.map(
                (v,i) =>
                  <li key={i}>
                    {v} <span onClick={() => onValueRemove(v)}>x</span>
                  </li>
              )
            }
          </div>
        }
      </div>
    )
  }
}

const styles = {
  wrapper: {
    display: 'inline-block'
  },
  values: {
    position: 'relative'
  },
  list: {
    display: 'none',
    listStyle: 'none',
    color: 'white'
  },
  listVisible: {
    display: 'inline-block'
  },
  listItem: {
    background: 'black',
    display: 'inline-block',
    cursor: 'pointer'
  },
  darth: {
    content: '',
    display: 'block',
    position: 'absolute',
    left: 0,
    top: 0,
    width: '100%',
    height: '100%',
    opacity: 0.6,
    backgroundImage: 'url(http://digitalspyuk.cdnds.net/15/50/1600x800/landscape-1449498579-darth-vader-star-wars.jpg)',
    backgroundSize: 'cover',
    backgroundPosition: 'center',
    backgroundRepeat: 'no-repeat',
    zIndex: '-1'
  }
}

export default onClickOutside(Radium(MultiSelect))
