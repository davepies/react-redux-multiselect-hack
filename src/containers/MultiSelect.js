import React from 'react'
import Select from '../components/Select'
import { connect } from 'react-redux'

import { showValues, hideValues, updateFilter } from '../state'

const MultiSelect = (props) => {
  return (
    <Select {...props} />
  )
}

export default connect(
  ({filter, values, filteredValues, showValues}) => ({
    values: filter ? filteredValues : values,
    filter,
    showValues
  }),
  (dispatch) => ({
    onFocus: () => { dispatch(showValues()) },
    onBlur: () => { dispatch(hideValues()) },
    onChange: (value) => { dispatch(updateFilter(value)) }
  })
)(MultiSelect)
