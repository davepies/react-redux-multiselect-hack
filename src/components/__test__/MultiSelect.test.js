/* globals describe, test, expect */

import React from 'react'
import MultiSelect from '../MultiSelect'

import { mount, render } from 'enzyme'

describe('Multiselect', () => {
  test('simple render', () => {
    const values = ['a', 'b', 'c']
    const multiSelectOutput = render(<MultiSelect values={values} />)

    expect(multiSelectOutput).toMatchSnapshot()
  })

  test('renders the same mount of list items as given values', () => {
    const listItemsLength = Math.ceil(Math.random() * 20)
    const values = Array(listItemsLength).fill('value')
    const multiSelectWrapper = mount(<MultiSelect values={values} />)

    const valuesLi = multiSelectWrapper.find('li')
    expect(valuesLi.length).toBe(values.length)
  })

  test('displays list of selectable values when focused', () => {
    const multiSelectWrapper = mount(<MultiSelect />)

    const valuesUl = multiSelectWrapper.find('ul').get(0)
    const filterInput = multiSelectWrapper.find('input')

    expect(valuesUl.props.style.display).toBe('none')

    filterInput.simulate('focus')
    const updatedValuesUl = multiSelectWrapper.find('ul').get(0)
    expect(updatedValuesUl.props.style.display).not.toBe('none')

    // expect(multiSelect).toMatchSnapshot()
    // manually trigger the callback
    // tree.props.onMouseEnter();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();

  })
})
