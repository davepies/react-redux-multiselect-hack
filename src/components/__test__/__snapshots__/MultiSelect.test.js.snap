// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiselect simple render 1`] = `
<div
  data-radium="true"
  style="display:inline-block"
>
  <input
    type="text"
  />
  <div
    data-radium="true"
    style="position:relative"
  >
    <ul
      data-radium="true"
      style="display:none;list-style:none;color:white"
    >
      <li
        data-radium="true"
      >
        <span
          data-radium="true"
          style="background:black;display:inline-block;cursor:pointer"
        >
          a
        </span>
      </li>
      <li
        data-radium="true"
      >
        <span
          data-radium="true"
          style="background:black;display:inline-block;cursor:pointer"
        >
          b
        </span>
      </li>
      <li
        data-radium="true"
      >
        <span
          data-radium="true"
          style="background:black;display:inline-block;cursor:pointer"
        >
          c
        </span>
      </li>
    </ul>
    <span
      data-radium="true"
      style="content:;display:block;position:absolute;left:0;top:0;width:100%;height:100%;opacity:0.6;background-image:url(http://digitalspyuk.cdnds.net/15/50/1600x800/landscape-1449498579-darth-vader-star-wars.jpg);background-size:cover;background-position:center;background-repeat:no-repeat;z-index:-1"
    />
  </div>
</div>
`;
